test/rules/no-unnecessary-type-assertion/test.ts.fix

```diff
@@ -69,3 +69,13 @@ function func(aOrB: TypeA|TypeB) {
         let z = aOrB;
     }
 }
+
+// Expecting no warning for these assertions as they are not unnecessary.
+
+type Bar = 'bar';
+const data = {
+    x: 'foo' as 'foo',
+    y: 'bar' as Bar,
```

Comment:
It's a relatively common pattern in our very large codebase where we're trying to turn on this rule. I don't think there's anything so wrong with the pattern that it would make sense to ban it. It keeps you from having to repeat the keys, at least.

Skimming through my list of breakages, I also see:

```typescript
const FOO = {
  bar: 'a' as 'a',
  baz: 'stuff',
  // ... many more properties without 'as'
```

i.e., not all properties are cast like this. For those I would hate to have to make it so much more verbose.

I also see instances of:

```typescript
for (const foo of ['bar' as 'bar', 'baz' as 'baz']) { /* ... */ }
```

Also, a search turned up a match in tslint itself:
https://github.com/palantir/tslint/blob/master/src/linter.ts#L110